Use a cast instead of trying to rely on higher level APIs.

gottesmm opened the pull request:
Use a cast instead of trying to rely on higher level APIs.

I missed this problem due to a missing test case in initializers.swift. In a
subsequent commit, I am going to add the rest of the missing cases with some
FileCheck tests for some of them. This is to ensure that the commit is easy to
review.

rdar://32539006
(cherry picked from commit cd55f466260b41b95cb0d113f9d7c6c86f6b0304)

gottesmm committed:
Use a cast instead of trying to rely on higher level APIs.

I missed this problem due to a missing test case in initializers.swift. In a
subsequent commit, I am going to add the rest of the missing cases with some
FileCheck tests for some of them. This is to ensure that the commit is easy to
review.

rdar://32539006
(cherry picked from commit cd55f466260b41b95cb0d113f9d7c6c86f6b0304)

gottesmm commented:
@swift-ci test

gottesmm commented:
@slavapestov Can you review?

slavapestov commented on lib/SILGen/SILGenExpr.cpp:
```diff
@@ -542,9 +542,24 @@ struct DelegateInitSelfWritebackCleanup : Cleanup {
+    // back into the self slot, we need to perform a downcast from the upcasted
+    // store value to the derived type of our lvalueAddress.
+    if (valueToStore->getType() != lvalueObjTy) {
+      if (!valueToStore->getType().isExactSuperclassOf(lvalueObjTy)) {
```
Could this be an assert instead?

gottesmm replied:
Sure.

slavapestov commented:
LGTM

gottesmm commented:
@spestov I fixed the issue.

gottesmm commented:
@swift-ci test

gottesmm commented:
@swift-ci test

gottesmm commented:
Explanation: This fixes an issue with delegating inits where an argument being thrown would result in self being casted incorrectly.  This got through our testing due to us not having tests that tested if we had an argument that could throw.
Scope: This is a relatively easy bug to hit if you do it the right way. So it is important to take this.
Radar (and possibly SR Issue): rdar://32539006
Risk: Minimal. This has baked over the week in tree where I added exhaustive tests (they are not in this specific commit to make reading the diff easy).
Testing: I added the missing test so just running the compiler tests should be sufficient.
